File: examples/python/06-oop-modules/main.py
"""
OOP and Modules in Python
========================
Demonstrates classes, objects, inheritance, encapsulation,
polymorphism, and modules (namespaces).
"""

# ── Classes and Objects ───────────────────────

print("--- Classes and Objects ---")

# Class definition
class Animal:
    def __init__(self, name: str, species: str):
        self.name = name
        self.species = species

    def speak(self) -> str:
        return f"{self.name} makes a sound"

    def __str__(self) -> str:
        return f"Animal({self.name}, {self.species})"


# Object instantiation
dog = Animal("Buddy", "Dog")
cat = Animal("Whiskers", "Cat")

print(f"dog = {dog}")
print(f"dog.speak() = {dog.speak()}")
print(f"str(dog) = {str(dog)}")

# Class vs Instance attributes
class Counter:
    count = 0  # Class attribute (shared)

    def __init__(self):
        # Reset instance attribute
        self.value = 0

    def increment(self):
        self.count += 1
        self.value += 1

counter1 = Counter()
counter2 = Counter()
counter1.increment()
counter2.increment()

print(f"counter1.count = {counter1.count} (class attribute)")
print(f"counter2.count = {counter2.count} (class attribute)")
print(f"counter1.value = {counter1.value} (instance attribute)")
print(f"counter2.value = {counter2.value} (instance attribute)")


# Instance methods vs Class methods
class InstanceMethods:
    def instance_method(self):
        return "Instance method"

    @classmethod
    def class_method(cls):
        return "Class method"

obj = InstanceMethods()
print(f"obj.instance_method() = {obj.instance_method()}")
print(f"obj.class_method() = {obj.class_method()}")


# ── Inheritance ───────────────────────────────

print("\n--- Inheritance ---")

# Base class
class Vehicle:
    def __init__(self, brand: str, year: int):
        self.brand = brand
        self.year = year

    def start(self) -> str:
        return f"{self.brand} engine starting"

    def stop(self) -> str:
        return f"{self.brand} engine stopping"


# Subclass with method override
class Car(Vehicle):
    def __init__(self, brand: str, year: int, model: str):
        super().__init__(brand, year)
        self.model = model

    def start(self) -> str:
        return f"{self.brand} {self.model} starting"

    def stop(self) -> str:
        return f"{self.brand} {self.model} stopping"


# Subclass with additional methods
class ElectricCar(Car):
    def __init__(self, brand: str, year: int, model: str, battery_capacity: int):
        super().__init__(brand, year, model)
        self.battery_capacity = battery_capacity

    def charge(self) -> str:
        return f"{self.brand} {self.model} charging"

    def get_range(self) -> int:
        return self.battery_capacity * 3


# Inheritance demonstration
tesla = ElectricCar("Tesla", 2020, "Model S", 400)
f150 = Car("Ford", 2020, "F-150", None)

print(f"tesla.start() = {tesla.start()}")
print(f"tesla.charge() = {tesla.charge()}")
print(f"tesla.get_range() = {tesla.get_range()} miles")

print(f"f150.start() = {f150.start()}")
print(f"f150.charge() = N/A (no battery)")
print(f"f150.get_range() = N/A (no battery)")

# isinstance and issubclass checks
print(f"tesla is a Vehicle? {isinstance(tesla, Vehicle)}")
print(f"tesla is a Car? {isinstance(tesla, Car)}")
print(f"ElectricCar is subclass of Car? {issubclass(ElectricCar, Car)}")


# ── Encapsulation ───────────────────────────────

print("\n--- Encapsulation ---")

# Private attributes (name mangling)
class BankAccount:
    def __init__(self, owner: str, balance: float):
        self.__owner = owner  # Private (name mangled)
        self.__balance = balance  # Private

    @property
    def owner(self) -> str:
        return self.__owner  # Public getter

    @property
    def balance(self) -> float:
        return self.__balance  # Public getter

    def deposit(self, amount: float) -> None:
        if amount > 0:
            self.__balance += amount

    def withdraw(self, amount: float) -> bool:
        if 0 < amount <= self.__balance:
            self.__balance -= amount
            return True
        return False


account = BankAccount("Alice", 1000.0)
print(f"account.owner = {account.owner}")
print(f"account.balance = {account.balance}")

account.deposit(500.0)
print(f"After deposit: {account.balance}")

account.withdraw(200.0)
print(f"After withdraw: {account.balance}")

# Direct attribute access (still possible with name mangling)
# account.__balance = 999999  # "Private" but accessible


# ── Polymorphism ─────────────────────────────

print("\n--- Polymorphism ---")

# Duck typing (Python's polymorphism style)
class Duck:
    def quack(self) -> str:
        return "Quack!"

    def swim(self) -> str:
        return "Duck swimming"


class Person:
    def quack(self) -> str:
        return "I'm quacking like a duck!"

    def swim(self) -> str:
        return "Person swimming (can't swim)"


class Robot:
    def quack(self) -> str:
        return "Beep boop!"

    def swim(self) -> str:
        return "Robot swimming!"


# Polymorphic function
def interact(entity):
    entity.quack()
    entity.swim()


print("Duck polymorphism:")
interact(Duck())    # Outputs: Quack! and Duck swimming

print("\nPerson polymorphism:")
interact(Person())  # Outputs: I'm quacking like a duck! and Person swimming (can't swim)

print("\nRobot polymorphism:")
interact(Robot())  # Outputs: Beep boop! and Robot swimming!


# ── Abstract Base Classes ─────────────────────

print("\n--- Abstract Base Classes (ABC) ---")

from abc import ABC, abstractmethod


class Shape(ABC):
    @abstractmethod
    def area(self) -> float:
        pass

    @abstractmethod
    def perimeter(self) -> float:
        pass


class Rectangle(Shape):
    def __init__(self, width: float, height: float):
        self.width = width
        self.height = height

    def area(self) -> float:
        return self.width * self.height

    def perimeter(self) -> float:
        return 2 * (self.width + self.height)


class Circle(Shape):
    def __init__(self, radius: float):
        self.radius = radius

    def area(self) -> float:
        import math
        return math.pi * self.radius ** 2

    def perimeter(self) -> float:
        return 2 * math.pi * self.radius


# Cannot instantiate abstract class
# shape = Shape()  # TypeError!

# Instantiate concrete subclasses
rectangle = Rectangle(5, 3)
circle = Circle(2)

print(f"rectangle.area() = {rectangle.area()}")
print(f"circle.area() = {circle.area()}")


# ── Multiple Inheritance ─────────────────────

print("\n--- Multiple Inheritance ---")


class Flyable:
    def fly(self) -> str:
        return "Flying!"


class Swimmable:
    def swim(self) -> str:
        return "Swimming!"


class Duck(Flyable, Swimmable):
    def __init__(self, name: str):
        self.name = name

    def fly(self) -> str:
        return f"{self.name} is flying!"

    def swim(self) -> str:
        return f"{self.name} is swimming!"

    def quack(self) -> str:
        return "Quack!"


class Penguin(Swimmable):
    def __init__(self, name: str):
        self.name = name

    def swim(self) -> str:
        return f"{self.name} is swimming!"

    def quack(self) -> str:
        return "No quack for penguin!"


# Multiple inheritance demonstration
duck = Duck("Donald")
penguin = Penguin("Penny")

print(f"duck.fly() = {duck.fly()}")
print(f"duck.swim() = {duck.swim()}")
print(f"duck.quack() = {duck.quack()}")

print(f"\npenguin.fly() = N/A (no Flyable)")
print(f"penguin.swim() = {penguin.swim()}")
print(f"penguin.quack() = {penguin.quack()}")


# ── Method Resolution Order ─────────────

print("\n--- Method Resolution Order (MRO) ---")


class A:
    def method(self) -> str:
        return "Method from A"


class B(A):
    def method(self) -> str:
        return f"Method from B (extends A)"


class C(B):
    def method(self) -> str:
        return f"Method from C (extends B)"


# Diamond problem
class D(B, C):
    def method(self) -> str:
        return f"Method from D (extends B and C)"


d = D()
print(f"d.method() = {d.method()}")
print(f"MRO of D: {[cls.__name__ for cls in D.__mro__]}")


# ── Properties ─────────────────────────────────────

print("\n--- Properties (getters/setters) ---")


class Temperature:
    def __init__(self, celsius: float):
        self._celsius = celsius

    @property
    def celsius(self) -> float:
        return self._celsius

    @celsius.setter
    def celsius(self, value: float):
        self._celsius = value

    @property
    def fahrenheit(self) -> float:
        return self._celsius * 9 / 5 + 32

    @fahrenheit.setter
    def fahrenheit(self, value: float):
        self._celsius = (value - 32) * 5 / 9


temp = Temperature()
temp.celsius = 25

print(f"temp.celsius = {temp.celsius}")
print(f"temp.fahrenheit = {temp.fahrenheit}")

temp.fahrenheit = 77
print(f"After setting F=77, C = {temp.celsius}")


# ── Static Methods and Class Methods ─────────

print("\n--- Static and Class Methods ---")


class Math:
    @staticmethod
    def add(a: float, b: float) -> float:
        return a + b

    @classmethod
    def multiply(cls, a: float, b: float) -> float:
        return a * b

    def divide(self, a: float, b: float) -> float:
        return a / b if b != 0 else 0.0


print(f"Math.add(3, 4) = {Math.add(3, 4)}")
print(f"Math.multiply(3, 4) = {Math.multiply(3, 4)}")
print(f"Math.divide(10, 2) = {Math.divide(10, 2)}")
print(f"Math.divide(10, 0) = {Math.divide(10, 0)}")


# ── Magic Methods ─────────────────────────────

print("\n--- Magic Methods (__dunder__) ---")


class Vector:
    def __init__(self, x: float, y: float):
        self.x = x
        self.y = y

    def __str__(self) -> str:
        return f"Vector({self.x}, {self.y})"

    def __repr__(self) -> str:
        return f"Vector(x={self.x}, y={self.y})"

    def __add__(self, other):
        return Vector(self.x + other.x, self.y + other.y)

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y


v1 = Vector(3, 4)
v2 = Vector(3, 4)
v3 = v1 + v2

print(f"v1 = {v1}")
print(f"v2 = {v2}")
print(f"v1 == v2: {v1 == v2}")
print(f"v1 + v2 = {v3}")
print(f"str(v1 + v2) = {str(v3)}")


# ── Dataclasses (Python 3.7+) ─────────────

print("\n--- Dataclasses (Python 3.7+) ---")

from dataclasses import dataclass


@dataclass
class Point:
    x: float
    y: float

    def distance_to(self, other: 'Point') -> float:
        return ((self.x - other.x) ** 2 + (self.y - other.y) ** 2) ** 0.5)


@dataclass
class PersonDataclass:
    name: str
    age: int
    email: str = ""


point = Point(0, 0)
person = PersonDataclass("Alice", 30)
print(f"point = {point}")
print(f"person = {person}")

# Dataclasses provide __init__, __repr__, __eq__ automatically


# ── Summary ───────────────────────────────────

print("\n--- Summary ---")
print("Python OOP:")
print("  - Classes: class keyword, __init__, self")
print("  - Objects: instantiate classes, dot notation")
print("  - Inheritance: extends base class, method override")
print("  - Multiple inheritance: comma-separated base classes")
print("  - Encapsulation: private (__attr) via name mangling")
print("  - Polymorphism: duck typing (no interfaces needed)")
print("  - Abstract base classes: ABC module, @abstractmethod")
print("  - Method Resolution Order: MRO for diamond problem")
print("  - Properties: @property decorator for getters/setters")
print("  - Static methods: @staticmethod, @classmethod")
print("  - Magic methods: __str__, __repr__, __eq__, etc.")
print("  - Dataclasses: type-safe data containers (3.7+)")

File: examples/typescript/06-oop-modules/main.ts
/*
 * OOP and Modules in TypeScript
 * =========================
 * Demonstrates classes, interfaces, inheritance, generics,
 * access modifiers, and modules. TypeScript adds static typing to JavaScript.
 */

// ── Classes and Objects ────────────────────────────

function demonstrateClassesObjects(): void {
  console.log("--- Classes and Objects ---");

  // Class definition
  class Animal {
    constructor(public name: string, public species: string) {}

    speak(): string {
      return `${this.name} makes a sound`;
    }
  }

  // Object instantiation
  const dog = new Animal("Buddy", "Dog");
  const cat = new Animal("Whiskers", "Cat");

  console.log(`dog = ${dog}`);
  console.log(`dog.speak() = ${dog.speak()}`);

  // Class vs Instance attributes
  class Counter {
    count: number = 0;  // Class attribute (shared)

    constructor() {
      // Reset instance attribute
      this.value = 0;
    }

    increment(): void {
      this.count++;
      this.value++;
    }
  }

  const counter1 = new Counter();
  const counter2 = new Counter();
  counter1.increment();
  counter2.increment();

  console.log(`counter1.count = ${counter1.count} (class attribute)`);
  console.log(`counter2.count = ${counter2.count} (class attribute)`);
  console.log(`counter1.value = ${counter1.value} (instance attribute)`);
  console.log(`counter2.value = ${counter2.value} (instance attribute)`);

  // Instance methods vs Class methods
  class InstanceMethods {
    instanceMethod(): string {
      return "Instance method";
    }

    static classMethod(): string {
      return "Class method";
    }
  }

  const obj = new InstanceMethods();
  console.log(`obj.instanceMethod() = ${obj.instanceMethod()}`);
  console.log(`obj.classMethod() = ${obj.classMethod()}`);
}


// ── Inheritance ───────────────────────────────

function demonstrateInheritance(): void {
  console.log("\n--- Inheritance ---");

  // Base class
  class Vehicle {
    constructor(public brand: string, public year: number) {}

    start(): string {
      return `${this.brand} engine starting`;
    }

    stop(): string {
      return `${this.brand} engine stopping`;
    }
  }

  // Subclass with method override
  class Car extends Vehicle {
    constructor(brand: string, year: number, private model: string) {}

    start(): string {
      return `${this.brand} ${this.model} starting`;
    }

    stop(): string {
      return `${this.brand} ${this.model} stopping`;
    }
  }

  // Subclass with additional methods
  class ElectricCar extends Car {
    constructor(brand: string, year: number, model: string, private batteryCapacity: number) {}

    charge(): string {
      return `${this.brand} ${this.model} charging`;
    }

    getRange(): number {
      return this.batteryCapacity * 3;
    }
  }

  // Inheritance demonstration
  const tesla = new ElectricCar("Tesla", 2020, "Model S", 400);
  const f150 = new Car("Ford", 2020, "F-150");

  console.log(`tesla.start() = ${tesla.start()}`);
  console.log(`tesla.charge() = ${tesla.charge()}`);
  console.log(`tesla.getRange() = ${tesla.getRange()} miles`);

  console.log(`f150.start() = ${f150.start()}`);
  console.log(`f150.charge() = N/A (no battery)`);
  console.log(`f150.getRange() = N/A (no battery)`);

  // instanceof and is subclass checks
  console.log(`tesla is a Vehicle? ${tesla instanceof Vehicle}`);
  console.log(`tesla is a Car? ${tesla instanceof Car}`);
  console.log(`ElectricCar is subclass of Car? ${tesla instanceof ElectricCar}`);

  // Protected access (TypeScript 3.8+)
  console.log(`tesla.model = ${tesla['model']}`);
}


// ── Access Modifiers ─────────────────────

function demonstrateAccessModifiers(): void {
  console.log("\n--- Access Modifiers ---");

  // Public (default)
  class PublicClass {
    public publicField: string = "public";
    public publicMethod(): string {
      return "public method";
    }
  }

  // Private
  class PrivateClass {
    private privateField: string = "private";
    private privateMethod(): string {
      return "private method";
    }
  }

  // Protected
  class ProtectedClass {
    protected protectedField: string = "protected";
    protected protectedMethod(): string {
      return "protected method";
    }

    // Read-only (readonly)
  class ReadonlyClass {
    readonly readonlyField: string = "readonly";

    public constructor(readonlyField: string) {
      this['readonlyField'] = readonlyField;
    }
  }

  const pubInst = new PublicClass();
  const privInst = new PrivateClass();
  const protInst = new ProtectedClass();

  console.log(`Public field: ${pubInst.publicField}`);
  console.log(`Private field (inaccessible): ${privInst['privateField']}`);
  console.log(`Protected field (accessible): ${protInst.protectedField}`);
  console.log(`Readonly field: ${pubInst['readonlyField']}`); // Can read, write via constructor

  // Static members
  class StaticMembers {
    public static staticField: string = "static field";
    public static staticMethod(): string {
      return "static method";
    }
  }

  console.log(`StaticMembers.staticField = ${StaticMembers.staticField}`);
  console.log(`StaticMembers.staticMethod() = ${StaticMembers.staticMethod()}`);

  const instance = new StaticMembers();
  console.log(`instance.staticField = ${instance['staticField']}`); // Can access static via instance
  console.log(`instance.staticMethod() = ${instance.staticMethod()}`);
}


// ── Abstract Classes ─────────────────────

function demonstrateAbstractClasses(): void {
  console.log("\n--- Abstract Classes ---");

  // Abstract class (cannot be instantiated)
  abstract class Shape {
    abstract area(): number;
    abstract perimeter(): number;
  }

  // Concrete class
  class Rectangle extends Shape {
    constructor(public width: number, public height: number) {}

    area(): number {
      return this.width * this.height;
    }

    perimeter(): number {
      return 2 * (this.width + this.height);
    }
  }

  // Cannot instantiate abstract class
  // const shape = new Shape();  // Compile error!

  // Instantiate concrete subclasses
  const rectangle = new Rectangle(5, 3);
  const circle = new Circle(2);

  console.log(`rectangle.area() = ${rectangle.area()}`);
  console.log(`circle.area() = ${circle.area()}`);
}


// ── Interfaces ─────────────────────────────

function demonstrateInterfaces(): void {
  console.log("\n--- Interfaces ---");

  // Interface definition
  interface Drawable {
    draw(): void;
    resize(factor: number): void;
  }

  // Class implementing interface
  class Circle implements Drawable {
    constructor(private radius: number) {}

    draw(): void {
      console.log(`Drawing circle with radius ${this.radius}`);
    }

    resize(factor: number): void {
      this.radius *= factor;
    }
  }

  class Rectangle implements Drawable {
    constructor(private width: number, private height: number) {}

    draw(): void {
      console.log(`Drawing rectangle ${this.width}x${this.height}`);
    }

    resize(factor: number): void {
      this.width *= factor;
      this.height *= factor;
    }
  }

  // Polymorphic usage
  const shape1: Drawable = new Circle(5);
  const shape2: Drawable = new Rectangle(10, 5);

  function drawAll(drawable: Drawable): void {
    drawable.draw();
  }

  drawAll(shape1);
  drawAll(shape2);

  // Multiple interfaces
  interface Swimmable {
    swim(): void;
  }

  interface Flyable {
    fly(): void;
  }

  class Duck implements Swimmable, Flyable {
    swim(): void {
      console.log("Duck swimming");
    }

    fly(): void {
      console.log("Duck flying!");
    }
  }

  const duck = new Duck();

  duck.swim();
  duck.fly();
}


// ── Properties (getters/setters) ─────────────

function demonstrateProperties(): void {
  console.log("\n--- Properties (getters/setters) ---");

  class Temperature {
    private _celsius: number;

    get celsius(): number {
      return this._celsius;
    }

    set celsius(value: number): void {
      this._celsius = value;
    }

    get fahrenheit(): number {
      return this._celsius * 9 / 5 + 32;
    }

    set fahrenheit(value: number): void {
      this._celsius = (value - 32) * 5 / 9;
    }
  }

  const temp = new Temperature();
  temp.celsius = 25;

  console.log(`temp.celsius = ${temp.celsius}`);
  console.log(`temp.fahrenheit = ${temp.fahrenheit}`);

  temp.fahrenheit = 77;
  console.log(`After setting F=77, C = ${temp.celsius}`);
}


// ── Generics ─────────────────────────────

function demonstrateGenerics(): void {
  console.log("\n--- Generics ---");

  // Generic class
  class Box {
    public value: T;

    constructor(value: T) {
      this.value = value;
    }
  }

  // Generic interface
  interface Repository {
    get(id: string): T;
    save(entity: T): void;
  }

  // Generic function
  function first<T>(array: T[]): T {
    return array[0];
  }

  const stringBox = new Box<string>("hello");
  const numberBox = new Box<number>(42);

  console.log(`stringBox.value = ${stringBox.value}`);
  console.log(`numberBox.value = ${numberBox.value}`);

  console.log(`first<string>(["a", "b"]) = ${first<string>(["a", "b"])}`);
  console.log(`first<number>([1, 2, 3]) = ${first<number>([1, 2, 3])}`);
}


// ── Method Parameters ─────────────────────

function demonstrateMethodParameters(): void {
  console.log("\n--- Method Parameters ---");

  // Optional parameters
  function greet(name: string, greeting?: string): string {
    return `${greeting}, ${name}!`;
  }

  // Default parameter values
  function multiply(a: number, b: number = 1): number {
    return a * b;
  }

  // Rest parameters
  function sum(...numbers: number[]): number {
    return numbers.reduce((acc, n) => acc + n, 0);
  }

  console.log(`greet("Alice") = ${greet("Alice")}`);
  console.log(`greet("Alice", "Hi") = ${greet("Alice", "Hi")}`);
  console.log(`multiply(5, 3) = ${multiply(5, 3)}`);
  console.log(`sum(1,2,3,4,5) = ${sum(1, 2, 3, 4, 5)}`);
}


// ── Enums ─────────────────────────────

function demonstrateEnums(): void {
  console.log("\n--- Enums ---");

  // Numeric enum
  enum Direction {
    Up,
    Down,
    Left,
    Right,
  }

  // String enum
  enum Color {
    Red,
    Green,
    Blue,
  }

  // Heterogeneous enum with values
  enum Status {
    Active = 200,
    Pending,
    Failed = { message: string, code: number },
  }

  // Enum usage
  const dir: Direction = Direction.Up;
  const color: Color = Color.Red;
  const status: Status.Status = Status.Active;

  console.log(`Direction: ${dir}`);
  console.log(`Color: ${color}`);
  console.log(`Status: ${status}`);

  // Enum members
  console.log(`All Directions: ${Object.values(Direction)}`);

  // Numeric enum values
  console.log(`Direction.Up value: ${Direction.Up}`);
}


// ── Decorators ───────────────────────────

function demonstrateDecorators(): void {
  console.log("\n--- Decorators ---");

  // Class decorator
  @sealed()
  class Animal {}

  @sealed()
  class Mammal extends Animal {}

  @sealed()
  class Reptile extends Mammal {}

  // Method decorator
  @log
  class Logger {
    @log
    private log(message: string): void {
      console.log(`[LOG] ${message}`);
    }
  }

  const logger = new Logger();

  // @deprecated
  class OldAPI {
    @deprecated("Use NewAPI instead")
    oldMethod(): string {
      return "old method";
    }

    newMethod(): string {
      return "new method";
    }
  }

  const api = new OldAPI();

  console.log("Using oldMethod (should warn):");
  api.oldMethod();

  console.log("Using newMethod (no warning):");
  api.newMethod();

  // Property decorator
  function readonly(target: any, key: string) {
    return {
      writable: false,
      configurable: false,
      get(): any {
        return target[key];
      }
    };
  }

  class Config {
    @readonly
    version: string = "1.0.0";

    @readonly
    apiUrl: string = "https://api.example.com";
  }

  const config = new Config();

  console.log(`config.version = ${config.version}`);
  console.log(`config.apiUrl = ${config.apiUrl}`);

  // config.version = "2.0.0";  // Error! readonly
}


// ── Modules (namespaces) ─────────────────

function demonstrateModules(): void {
  console.log("\n--- Modules (namespaces) ---");

  // Export default
  export class PublicClass {
    public value: string = "exported class";
  }

  export const constant = "exported constant";

  // Import named
  import { Logger } from "./logger";

  // Import all (wildcard)
  import * as utils from "./utils";

  // Default export
  export default class DefaultExport {
    public value: string = "default export";
  }

  // Namespace export
  export namespace MyNamespace {
    export class NamespacedClass {
      public value: string = "namespaced class";
    }
  }

  // Re-export
  export { PublicClass, constant, DefaultExport } from "./re-export";

  // Type-only imports
  import type { Animal } from "./types";

  // Ambient modules
  declare const GlobalValue: string;

  console.log(`GlobalValue = ${GlobalValue}`);
}


// ── Summary ───────────────────────────────────

function main(): void {
  console.log("OOP and Modules in TypeScript");
  console.log("==========================");

  demonstrateClassesObjects();
  demonstrateInheritance();
  demonstrateAccessModifiers();
  demonstrateAbstractClasses();
  demonstrateInterfaces();
  demonstrateProperties();
  demonstrateGenerics();
  demonstrateMethodParameters();
  demonstrateEnums();
  demonstrateDecorators();
  demonstrateModules();

  console.log("\n--- Summary ---");
  console.log("TypeScript OOP:");
  console.log("  - Classes: class keyword, constructor, public/private");
  console.log("  - Inheritance: extends keyword, method override");
  console.log("  - Multiple inheritance: no support (use composition)");
  console.log("  - Access modifiers: public, private, protected, readonly");
  console.log("  - Abstract classes: abstract keyword, abstract methods");
  console.log("  - Interfaces: contracts with implements keyword");
  console.log("  - Generics: <T> type parameters");
  console.log("  - Enums: enum keyword with values");
  console.log("  - Decorators: @sealed, @log, @deprecated");
  console.log("  - Modules: export/import, namespaces");
  console.log("  - Type inference: automatic with optional static typing");
}
